### PR TITLE
NXP-29580: Runtime Descriptors Migrations

### PR DESCRIPTION
**Pull request by atchertchian**

### CODE REVIEW

File: modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/impl/ExtensionPointImpl.java
Comment (efge) [inline modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/impl/ExtensionPointImpl.java:93]: You don't need `i`, so use `for (Class<?> contrib : contributions) {`

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/Registry.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/Registry.java:38]: Naming suggestion: `tag` or `label` instead of `mark`.

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/Registry.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/Registry.java:36]: You should mention that a registry can have not just zero or one but also more than one.

File: modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/DummyRegistry.java
Comment (efge) [inline modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/DummyRegistry.java:42]: `flag` -> `mark` (or whatever new name is chosen)

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedMembers.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedMembers.java:54]: XAnnotatedMember[] members = Arrays.stream(values)
                                       .map(value -> new XAnnotatedReference(xmap, String.class, value, null, null))
                                       .toArray(XAnnotatedMember[]::new);

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedMembers.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedMembers.java:79]: return Arrrays.stream(members).anyMatch(member -> member.hasValue(ctx, element));

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedMembers.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedMembers.java:94]: `return String.join(separator, values);`

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedMembers.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedMembers.java:101]: `Arrays.asList(members).forEach(member -> member.toXML(instance, parent))`

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XMap.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XMap.java:280]: `xregistryId.value()` can be put in a local variable

File: modules/runtime/nuxeo-runtime/src/test/java/org/nuxeo/runtime/TestDescriptorRegistry.java
Comment (efge) [inline modules/runtime/nuxeo-runtime/src/test/java/org/nuxeo/runtime/TestDescriptorRegistry.java:33]: `**/` -> `*/`

File: modules/runtime/nuxeo-runtime-cluster/src/main/java/org/nuxeo/runtime/cluster/ClusterServiceImpl.java
Comment (efge) [inline modules/runtime/nuxeo-runtime-cluster/src/main/java/org/nuxeo/runtime/cluster/ClusterServiceImpl.java:69]: Is the cast needed?

File: modules/runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java
Comment (efge) [inline modules/runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java:200]: Cast needed?

File: modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/services/config/TestConfigurationService.java
Comment (efge) [inline modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/services/config/TestConfigurationService.java:123]: Do you want to keep this comment?

File: modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/services/config/ConfigurationServiceImpl.java
Comment (efge) [inline modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/services/config/ConfigurationServiceImpl.java:152]: This seems weird. Isn't this simpler?

    desc.map(value -> value.getValue()).filter(StringUtils::isNotBlank);

File: modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/services/config/ConfigurationServiceRegistry.java
Comment (efge) [inline modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/services/config/ConfigurationServiceRegistry.java:30]: `@since 11.5`

File: modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/registry/SampleComponent.java
Comment (efge) [inline modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/registry/SampleComponent.java:35]: To follow other usage I would have named these `SINGLE_XP` etc.

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/annotation/XNode.java
Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/annotation/XNode.java:52]: I'm actually not sure as the expression is not "real" xpath, hence "xpath-like" or "xpathy", i can rephrase in all cases.

File: modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/ExtensionPoint.java
Comment (atchertchian) [reply modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/ExtensionPoint.java:76]: This depends if we'd like to keep ExtensionPointImpl as a simple POJO.
Also using Optional might interfere with existing serialization logic.

File: modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/impl/ComponentManagerImpl.java
Comment (atchertchian) [reply modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/impl/ComponentManagerImpl.java:583]: I changed it, please tell if still not that-readable

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/Registry.java
Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/Registry.java:38]: Used "tag" instead

File: modules/runtime/nuxeo-runtime-cluster/src/main/java/org/nuxeo/runtime/cluster/ClusterServiceImpl.java
Comment (atchertchian) [reply modules/runtime/nuxeo-runtime-cluster/src/main/java/org/nuxeo/runtime/cluster/ClusterServiceImpl.java:69]: Yes: "Type mismatch: cannot convert from Object to ClusterNodeDescriptor"

File: modules/runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java
Comment (atchertchian) [reply modules/runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java:200]: Same: orElse brings ambiguity hence the cast

File: modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/services/config/TestConfigurationService.java
Comment (atchertchian) [reply modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/services/config/TestConfigurationService.java:123]: no :)

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java
Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java:63]: Not sure indeed, to be checked

**Comment by atchertchian** [comment]
Thanks for the feedback, here are some TODO notes for this PR:
- discuss + check unresolved conversations
- introduce compat behavior for merge (?) unexpected merges could be turned down for compatibilty using a runtime property (?)
- migrate remaining Descriptor implementations before migrating all other descriptors

File: modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/services/config/ConfigurationServiceRegistry.java
Comment (atchertchian) [inline modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/services/config/ConfigurationServiceRegistry.java:48]: Note i made the "initialize" method public on registry, and explicitly called on component start, so that these deprecation logs can be shown at framework startup

**Comment by atchertchian** [comment]
There are some remaining Descritptor implementations but they're a bit more complicated to handle or have interactions with other contributions, not yet using registries, so they will be dealt with in later PRs.

Update on the TODO list:
* discuss + check unresolved conversations (about multi-threaded usage mostly for now)
* check tests status

Compatibility behavior for merge has been discarded for now (not easy to turn off merge conditionally in annotations), this will encourage migration, even if a list of impacted extension points could be useful for migration.

File: modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/services/config/ConfigurationServiceRegistry.java
Comment (atchertchian) [reply modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/services/config/ConfigurationServiceRegistry.java:48]: (Refined later when working on automation registries, to be able to extract the extension id from tag)

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/annotation/XNode.java
Comment (efge) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/annotation/XNode.java:52]: Not very important, let's keep as is.

File: modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/ExtensionPoint.java
Comment (efge) [reply modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/ExtensionPoint.java:76]: Ok let's keeps is simple and old-style.

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java
Comment (efge) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java:63]: The registration part seems safe as it's called by the startup thread. However other `getContribution*` calls are potentially done after startup.
So let's play it safe: make the `initialized` flag volatile and use double-checked locking to change it (and also setting it in `setInitialized` should also have a synchronization).
Also the datastructures in the registries (`tags`, `registrations` in the abstract class and other ones in the concrete classes) should all be thread-safe (—> `new ConcurentHashMap`, `ConcurrentHashMap.newKeySet()`, `new CopyOnWriteArrayList`, and read/write the `enabled` flag in a synchronized block too)

Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java:67]: Pushed in a separate commit for now, please let me know if that matches the need.

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedObject.java
Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XAnnotatedObject.java]: I did not see any need of extracting the cause in runtime unit tests so i changed this in a separate commit, to see how it goes with all other tests.

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/SingleRegistry.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/SingleRegistry.java:35]: `contribution` must be volatile too, as it's read by `getContribution()` which is potentially called from different threads.

**Review by efge — CHANGES_REQUESTED** [review]

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java
Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java:64]: Yes but i see no reason of preventing contributions retrieval (so registry initialization) before the framework triggers it when starting.

Comment (efge) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java:64]: Lazy init (on-demand) is always more complex, as we see. But that should be ok with the changes you have.

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java
Comment (efge) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java:87]: Ah good point, the concurrent maps disallow null keys...

**Review by efge — DISMISSED** [review]

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java
Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java:87]: This was visible in tests (core bulk config declares a logConfig without id, maybe this should be changed)

**Review by kevinleturc — APPROVED** [review]
Very nice work 👌.

Just details/suggestions that can be addressed in the future.

Indeed, the automatic type inference is a pain, mainly for Optional because we would like to chain the result directly.
See below my thoughts for future improvements:
- replace cast by manual type inference
- add back APIs with the class parameter, in addition to current APIs
- add a getNullableRegistryContribution for former usages

Let's merge it 💪

**Comment by atchertchian** [comment]
Squashed a few commits (and took latest feedback into account, that should not impact tests), waiting for a last complete validation before merge.

**Review by efge — APPROVED** [review]

**Review by kevinleturc — APPROVED** [review]

File: modules/platform/nuxeo-template-rendering/nuxeo-template-rendering-core/src/main/java/org/nuxeo/template/serializer/service/TemplateSerializerServiceImpl.java
Comment (atchertchian) [reply modules/platform/nuxeo-template-rendering/nuxeo-template-rendering-core/src/main/java/org/nuxeo/template/serializer/service/TemplateSerializerServiceImpl.java:56]: Done with #4575

File: modules/platform/nuxeo-automation/nuxeo-automation-server/src/main/java/org/nuxeo/ecm/automation/server/AutomationServerComponent.java
Comment (atchertchian) [reply modules/platform/nuxeo-automation/nuxeo-automation-server/src/main/java/org/nuxeo/ecm/automation/server/AutomationServerComponent.java:62]: Done with #4575

File: modules/core/nuxeo-core-cache/src/main/java/org/nuxeo/ecm/core/transientstore/TransientStorageComponent.java
Comment (atchertchian) [reply modules/core/nuxeo-core-cache/src/main/java/org/nuxeo/ecm/core/transientstore/TransientStorageComponent.java:111]: Done with #4575

**Pull request merged**
